src/auth/auth.service.ts

```diff
@@ -0,0 +1,185 @@
+/*
+ * SPDX-FileCopyrightText: 2021 The HedgeDoc developers (see AUTHORS file)
+ *
+ * SPDX-License-Identifier: AGPL-3.0-only
+ */
+
+import { Injectable } from '@nestjs/common';
+import { UsersService } from '../users/users.service';
+import { User } from '../users/user.entity';
+import { AuthToken } from './auth-token.entity';
+import { AuthTokenDto } from './auth-token.dto';
+import { AuthTokenWithSecretDto } from './auth-token-with-secret.dto';
+import { compare, hash } from 'bcrypt';
+import { NotInDBError, TokenNotValidError } from '../errors/errors';
+import { randomBytes } from 'crypto';
+import { InjectRepository } from '@nestjs/typeorm';
+import { Repository } from 'typeorm';
+import { ConsoleLoggerService } from '../logger/console-logger.service';
+import { TimestampMillis } from '../utils/timestamp';
+
+@Injectable()
+export class AuthService {
+  constructor(
+    private readonly logger: ConsoleLoggerService,
+    private usersService: UsersService,
+    @InjectRepository(AuthToken)
+    private authTokenRepository: Repository<AuthToken>,
+  ) {
+    this.logger.setContext(AuthService.name);
+  }
+
+  async validateToken(token: string): Promise<User> {
+    const [keyId, secret] = token.split('.');
+    const accessToken = await this.getAuthTokenAndValidate(keyId, secret);
+    const user = await this.usersService.getUserByUsername(
+      accessToken.user.userName,
+    );
+    if (user) {
+      await this.setLastUsedToken(keyId);
```

Comment:
Why do we only run this in case the previous query resulted in a user? 

Also I don't think we need to await this call since even if there is a race condition, the difference in time is basically meaningless.

Reply:
This was necessary before, but now isn't necessary anymore. I moved the call above the user search.
It should be fixed now.